src/fromager/vendor_rust.py

```diff
@@ -0,0 +1,192 @@
+
+
+def _tar_reset(tarinfo: tarfile.TarInfo) -> tarfile.TarInfo:
+    """Reset user, group, mtime, and mode to create reproducible tar"""
+    tarinfo.uid = 0
+    tarinfo.gid = 0
+    tarinfo.uname = "root"
+    tarinfo.gname = "root"
+    tarinfo.mtime = 0
+    if tarinfo.type == tarfile.DIRTYPE:
+        tarinfo.mode = 0o755
+    else:
+        tarinfo.mode = 0o644
+    if tarinfo.pax_headers:
+        raise ValueError(tarinfo.name, tarinfo.pax_headers)
+    return tarinfo
+
+
+def _tar_reproducible(tar: tarfile.TarFile, basedir: pathlib.Path) -> None:
+    """Create reproducible tar file"""
+    content = []
+    for root, dirs, files in basedir.walk():
+        for directory in dirs:
+            content.append(root / directory)
+        for filename in files:
+            content.append(root / filename)
+
+    content.sort()
+
+    for fn in content:
+        tar.add(
+            fn,
+            filter=_tar_reset,
+            recursive=False,
+            arcname=fn.relative_to(basedir),
+        )
+
+
+def _cargo_vendor(
+    manifests: typing.List[pathlib.Path],
+    project_dir: pathlib.Path,
+) -> None:
+    """Run cargo vendor"""
+    args = ["cargo", "vendor", f"--manifest-path={manifests[0]}"]
+    for manifest in manifests[1:]:
+        args.append(f"--sync={manifest}")
+    args.append(project_dir / VENDOR_DIR)
+    subprocess.check_call(args, stdout=subprocess.DEVNULL, env={})
+
+    # Remove large, unused archive and lib files. They are only
+    # used on Windows, macOS, and iOS. This makes the vendor bundle up to
+    # 60% smaller.
+    for pattern in VENDOR_FILTER_PATTERNS:
+        for filename in project_dir.glob(pattern):
+            filename.unlink()
+            filename.touch()  # create empty file
+
+
+def _fixup_cargo_manifest(
+    fileame: pathlib.Path,
+    *,
+    msrv: typing.Optional[Version] = None,
+):
+    """Modify Cargo.toml
+
+    pydantic-core has an unnecessary high MSRV.
+    """
+    with open(fileame, "r", encoding="utf-8") as f:
+        cfg = toml.load(f)
+
+    modified = False
+    if msrv is not None:
+        packaging = cfg.get("packaging", {})
+        rust_version = packaging.get("rust-version")
+        if rust_version is not None and Version(rust_version) >= msrv:
+            packaging["rust-version"] = str(msrv)
+            modified = True
+
+    if modified:
+        with open(fileame, "w", encoding="utf-8") as f:
+            toml.dump(cfg, f)
+
+
+def _cargo_config(project_dir: pathlib.Path):
+    """create .cargo/config.toml"""
+    dotcargo = project_dir / ".cargo"
+    config_toml = dotcargo / "config.toml"
+    try:
+        with open(config_toml, "r", encoding="utf-8") as f:
+            cfg = toml.load(f)
+    except FileNotFoundError:
+        dotcargo.mkdir(exist_ok=True)
+        cfg = {}
+
+    cfg.update(CARGO_CONFIG)
+
+    with open(config_toml, "w", encoding="utf-8") as f:
+        toml.dump(cfg, f)
+
+
+def vendor_rust(
+    sdist: pathlib.Path,
```

Comment:
We should work out how to integrate this into the flow.

Today we have

- download-source
- prepare-source
  - unpack
  - patch
- prepare-build
- build-wheel

That leaves the patched source tree with build artifacts in it on the disk, and leaves the original sdist unmodified.

I think we want instead something like

- download-source
- prepare-source
  - unpack
  - patch
  - vendor rust
- re-roll-sdist (to preserve a copy of our patched sources with the vendored rust code, before any build artifacts are written, so we can distribute the sources we use for the build)
- prepare-build
- build-wheel

If that's the flow, then I think for input we'll want this function to take the path to the source root instead of the sdist file.